types/sinon/index.d.ts

```diff
@@ -19,13 +19,17 @@ interface Event {} // tslint:disable-line no-empty-interface
 interface Document {} // tslint:disable-line no-empty-interface
 
 declare namespace Sinon {
-    type DeepPartialOrMatcher<T> = {
-        [K in keyof T]?: SinonMatcher | (T[K] extends object ? DeepPartialOrMatcher<T[K]> : T[K]);
+    type MatchProvidedArguments<T> = {
+        [K in keyof T]?: SinonMatcher | (T[K] extends object ? MatchProvidedArguments<T[K]> : T[K]);
     };
+    // Alias for backward compatibility, remove on next major release
+    type DeepPartialOrMatcher<T> = MatchProvidedArguments<T>;
 
-    type MatchArguments<T> = {
-        [K in keyof T]: SinonMatcher | (T[K] extends object ? MatchArguments<T[K]> : never) | T[K];
+    type MatchExactlyArguments<T> = {
```

Comment:
personally i would've gone with `MatchArguments` and `MatchExactArguments`

since thats what the sinon methods do: match the arguments vs. matching the exact arguments (i.e. no extras)

but if there's good reason to not go with that naming, im open to leaving it like it is

Reply:
I thought the same. However, these types are also public, so I was afraid of changing expectations on consumers. If we we upgrade major with this change I can leave the naming as you mentioned.

Reply:
Could we maybe go with `MatchPartialArguments` and `MatchExactArguments`?

Otherwise can leave it as is. You have a good point about backwards compat

Reply:
Done. I left the aliases for backwards compatibility.